### PR TITLE
[SPARK-4122][STREAMING] Add a library that can write data back to Kafka ...

### PR DESCRIPTION
**Pull request by harishreedharan**
...from Spark Streaming.

This adds a library that can writes dstreams to Kafka. An implicit also has been added so users
can call dstream.writeToKafka(..)

### CODE REVIEW

**Comment by harishreedharan** [comment]
Sure. I will work on a short design doc soon

**Comment by harishreedharan** [comment]
Here is a design doc: https://docs.google.com/document/d/1jy160SNSJyrkck7rM9IFrFxa3OFQbPwv3S-pH0rGyjU/edit?usp=sharing

**Comment by tdas** [comment]
@harishreedharan Thanks for the design doc. Can you make it open for public to comment :)

**Comment by harishreedharan** [comment]
Done!

**Comment by BenFradet** [comment]
@tdas @harishreedharan Any updates on this?

Since we're incorporating [Kafka 0.8.2.1](https://github.com/apache/spark/pull/4537) and that there is a [new Producer API](http://kafka.apache.org/082/javadoc/index.html?org/apache/kafka/clients/producer/KafkaProducer.html), there might be a need to start over.

It might be interesting to think about pooling producers (as well as consumers, for that matter) also.

**Comment by koeninger** [comment]
I have a branch of the directStream api that caches consumers.

It had no noticeable impact on processing time.   Even at 100 partitions /
200ms batches on a production-like workload.

**Comment by BenFradet** [comment]
Interesting, thanks a lot for sharing your findings.
Any info regarding producers?

**Comment by harishreedharan** [comment]
I have a version of this here: https://github.com/harishreedharan/spark-streaming-kafka-output

I am not sure this is going to make it to core spark, but you can use that one. At some point, I will get this populated to the cloudera mvn repo for ease of use.

**Comment by tdas** [comment]
@BenFradet  That's a good point. This is dropped off my radar but I really would love to have this in. @harishreedharan Let;s plan to get this in 1.5.0 release. Ping me after 1.4.0 release so that it does not fall off my radar again. :)

**Comment by harishreedharan** [comment]
If there is interest, I might change the code to do writes from an RDD and use the RDD as the basis for the stream (a la Direct Kafka Stream - @koeninger route)

**Comment by tdas** [comment]
THAT IS A NEAT IDEA. We should have bindings for both RDD and DStream.

**Comment by BenFradet** [comment]
@harishreedharan Thanks a lot for sharing, I will have to test that.

**Comment by 91pavan** [comment]
@harishreedharan Any chance I can use your version from the Spark Java API?

**Comment by harishreedharan** [comment]
I have not tested it. Go ahead and try it out and let me know.

On Friday, May 1, 2015, Pavan Sudheendra <notifications@github.com> wrote:

> @harishreedharan <https://github.com/harishreedharan> Any chance I can
> use your version from the Spark Java API?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/2994#issuecomment-98316465>.
>


-- 

Thanks,
Hari

**Comment by BenFradet** [comment]
@tdas

Also, as we were discussing with @harishreedharan, there is the issue of broadcasting the producer config.

Indeed, if we restore our streaming context from a checkpoint with ```getOrCreate``` we'll get an exception when trying to use the broadcast variable as broadcast variables are not checkpointed, this has already come up [on the mailing list](http://apache-spark-user-list.1001560.n3.nabble.com/NullPointerException-on-reading-checkpoint-files-td7306.html).

Maybe we should open a separate jira for this as this seems to be a valid use case afaic.

**Comment by tdas** [comment]
The best way to use a broadcast variable is to make a transient lazy
singleton.
See the way I suggested for Accumulators in
https://issues.apache.org/jira/browse/SPARK-5206

**Comment by BenFradet** [comment]
Thanks for the advice.

**Comment by harishreedharan** [comment]
I am planning to refactor this quite a bit anyway. I plan to update this with an RDD friendly version which can be used to implement one for DStreams too. Once I get to working on this, I will update this.

**Comment by BenFradet** [comment]
@tdas 

I was thinking we should maybe document this use case in [the checkpointing section of the streaming programming guide](http://spark.apache.org/docs/latest/streaming-programming-guide.html#checkpointing), since it's basically the same issue with broadcast variables, accumulators and external RDDs.
What do you think?

**Comment by tdas** [comment]
Yes, I plan to do so when I update the documentation for Spark 1.4 soon.

**Comment by BenFradet** [comment]
Great!

**Comment by harishreedharan** [comment]
I will get around to this in a few days.

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaOutputWriter.scala [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    val broadcastedConfig = dstream.ssc.sc.broadcast(producerConfig)
+
+    def func = (rdd: RDD[T]) => {
+      rdd.foreachPartition(events => {
+        // The ForEachDStream runs the function locally on the driver.
+        // This code can alternatively use sc.runJob, but this approach seemed cleaner.
+        val producer: Producer[K, V] =
+          new Producer[K, V](new ProducerConfig(broadcastedConfig.value))
+        try {
+          producer.send(events.map(serializerFunc).toArray: _*)
```
Comment (jerryshao) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaOutputWriter.scala]: Hi @harishreedharan , the Kafka producer API is changed after 0.8.2, you may also need to update the code accordingly.

Comment (harishreedharan) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaOutputWriter.scala]: I am not really updating this branch anymore. I am maintaining it here: https://github.com/cloudera/spark-kafka-writer

**Comment by andrewor14** [comment]
@harishreedharan any updates on this? Could you rebase to master?

**Comment by harishreedharan** [comment]
Sure, if there is interest in this. I have a version of this on the cloudera github (which I pushed there, since I didn't see much interest in having this in spark itself). I will port it over.

**Comment by WangTaoTheTonic** [comment]
Hi @harishreedharan, I think this is a nice feature which is very helpful for user who tries to write DStream back into Kafka. The implement is very neat too.

Could you please rebase the code so that we have the opptunity to push it into spark?

**Comment by harishreedharan** [comment]
I will get back to this one in a week or so

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaOutputWriter.scala [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    serializerFunc: T => KeyedMessage[K, V]): Unit = {
+
+    // Broadcast the producer to avoid sending it every time.
+    val broadcastedConfig = dstream.ssc.sc.broadcast(producerConfig)
+
+    def func = (rdd: RDD[T]) => {
+      rdd.foreachPartition(events => {
+        // The ForEachDStream runs the function locally on the driver.
+        // This code can alternatively use sc.runJob, but this approach seemed cleaner.
+        val producer: Producer[K, V] =
```
Comment (mkuthan) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaOutputWriter.scala]: You could consider share kafkaproducer per JVM/executor as long as producer is thread safe. I'm affraid that new producer for every partition will be main bottleneck in real-word scenarios.
Please refer to this blog post: http://allegro.tech/2015/08/spark-kafka-integration.html

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaOutputWriter.scala (hunk 1 above)
Comment (mkuthan) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaOutputWriter.scala]: Please remember that kafka producer is asynchronous. The processing should wait until the results are really published to the kafka cluster.  Please check how it is implemented in Samza:
https://github.com/apache/samza/blob/master/samza-kafka/src/main/scala/org/apache/samza/system/kafka/KafkaSystemProducer.scala

I'm not sure that we need retry policy, but at least we should check the result of the producer.send.

My $.02 cents :-)

**Comment by zzcclp** [comment]
@harishreedharan , dose this pr release with 1.6?

**Comment by harishreedharan** [comment]
I am closing this PR. I have put this up here: https://github.com/cloudera/spark-kafka-writer

**Pull request closed without merging**